gpbft/emulator/emulator.go

```diff
@@ -0,0 +1,153 @@
+// allows frame-by-frame control over progress of a GPBFT instance.
+type Emulator struct {
+	gpbft.Signatures
+
+	signing *signing.FakeBackend
+	require *require.Assertions
+	chain
+	id      gpbft.ActorID
+	subject *gpbft.Participant
+	now     time.Time
+	bus     eventBus
+
+	lastSubjectBroadcast *gpbft.GMessage
+	rng                  *rand.Rand
+}
+
+func New(t *testing.T, o ...gpbft.Option) *Emulator {
+	sb := signing.NewFakeBackend()
+	e := &Emulator{
+		require:    require.New(t),
+		rng:        rand.New(rand.NewSource(1413)), // TODO: parameterise
+		chain:      newChain(),
+		signing:    sb,
+		Signatures: sb,
+	}
+	participant, err := gpbft.NewParticipant(e, o...)
+	require.NoError(t, err)
+	e.subject = participant
+	return e
+}
+
+func (e *Emulator) SetInstance(instance *Instance) {
+	if !instance.PowerTable.Has(e.id) {
+		e.require.Fail("unknown subject storage power", "power table must contain storage power for emulated subject with ID: %d", e.id)
+	}
+	for i, entry := range instance.PowerTable.Entries {
+		// TODO: refactor signing backends to top level package and standardise the APIs
+		//       for better UX.
+		instance.PowerTable.Entries[i].PubKey = e.signing.Allow(int(entry.ID))
+	}
+	e.require.NoError(e.chain.setInstance(instance))
+}
+
+func (e *Emulator) StartInstance(id uint64) {
+	instance := e.getInstance(id)
+	e.require.NotNil(instance, "instance ID %d not found", id)
+	e.require.NoError(e.subject.StartInstance(id))
+}
+
+func (e *Emulator) NetworkName() gpbft.NetworkName { return "emulator-net" }
+
+func (e *Emulator) RequestBroadcast(mb *gpbft.MessageBuilder) error {
+	mb.SetNetworkName(e.NetworkName())
+	mb.SetSigningMarshaler(e)
+	msg, err := mb.Build(e.signing, e.id)
+	if err != nil {
+		return err
+	}
+	e.bus.enqueueBroadcast(msg)
+	return nil
+}
+
+func (e *Emulator) RequestBroadcastFrom(from gpbft.ActorID, mb *gpbft.MessageBuilder) {
+	e.require.NotEqual(e.id, from, "broadcast request on behalf of subject is not allowed")
+	mb.SetNetworkName(e.NetworkName())
+	mb.SetSigningMarshaler(e)
+	msg, err := mb.Build(e.signing, from)
+	e.require.NoError(err)
+	e.bus.enqueueBroadcast(msg)
+}
+
+func (e *Emulator) doBroadcast(msg *gpbft.GMessage) error {
+	if validated, err := e.subject.ValidateMessage(msg); err != nil {
+		return err
+	} else {
+		return e.subject.ReceiveMessage(validated)
+	}
+}
+
+func (e *Emulator) ReceiveDecision(decision *gpbft.Justification) time.Time {
+	e.require.NoError(e.chain.setDecision(decision))
+	e.now = e.Time().Add(1) // TODO: do we need to increment?
+	return e.now
+}
+
+func (e *Emulator) Time() time.Time { return e.now }
+
+func (e *Emulator) SetAlarm(at time.Time) {
+	e.bus.enqueueAlarm(&alarm{at: at})
+}
+
+func (e *Emulator) HasNext() bool {
+	return e.bus.Len() > 0
+}
+
+func (e *Emulator) NextUntil(match Matcher, timeout time.Duration, msgAndArgs ...any) {
+	if match() {
+		return
+	}
+
+	timer := time.NewTimer(timeout)
+	defer func() { timer.Stop() }()
```

Comment:
I think it primarily prevents a no-progress from stopping the test execution without relying on the global test timeout.

Reply:
@anorth it asserts that the condition occurs within a given timeout. 

> a worrying sign

Why? go tests have a default timeout of 10 minutes for a test to pass. There should be _some_ timeout when testing anything. Here, the code asserts that a condition becomes true within a given amount of time.

Reply:
We discussed this in person, and the main reason it was worrying to me is because of what it implied about how the tests were run: non-deterministically. I think we've agreed to change that.

But for posterity I'll note here that I would still try to avoid timeouts and make the tests clearly pass or fail. Is it _really_ any number of messages before the condition is met? Can we not deterministically detect a lack of progress? The simulator-based tests have far less control over how the participant are driven yet they don't need a timeout.

A timeout either signals that a test is flaky, or makes it so if it wasn't already. It's an assertion about hardware and environment. Any timeout low enough to be useful (i.e. signal "we broke something if this is exceeded") is also low enough to be accidentally exceeded in the wrong environment. So even if the test was deterministic, the timeout adds a spurious failure mode. Go's 10-minute default has some utility in preventing an infinite loop from locking up your CI machine for hours or days, but it's set to a ridiculous high value. I wouldn't have objected so much to a 10-minute timeout here either. But 2 seconds appears to be attempting to set an actual expectation about how long the code should run for. I don't think one should ever try to code "test should finish in 2 seconds, or it is a failure". (Benchmarks are a different thing, and use benchmark tools, not timeouts either).

Reply:
Thank you for capturing this, Alex.

I agree with the sentiment of the comment above. However, I continue to believe that in this case timeout is relevant. It is a scaled down version of the Go's default 10m timeout, offering sufficient headroom for me to be confident that it will not impact the test flakiness. Because, `Next` is sufficiently fast to result in thousands of iterations (i.e. broadcasts or alarm triggers) per second, and that means it should be a long enough timeout in any reasonable environment despite the "2s" itself being a short period of time. 

Based on what we discussed this timeout will go away from the PR in any case since the automatic dispatch of events via bus in random order will be removed from the PR.